src/Dependencies/PooledObjects/PooledDelegates.cs

```diff
@@ -0,0 +1,155 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.CodeAnalysis.PooledObjects
+{
+    internal static class PooledDelegates
+    {
+        private static class DefaultDelegatePool<T>
+            where T : class, new()
+        {
+            public static readonly ObjectPool<T> Instance = new ObjectPool<T>(() => new T(), 20);
+        }
+
+        private static Releaser GetPooledDelegate<TPooled, TArg, TUnboundDelegate, TBoundDelegate>(TUnboundDelegate unboundDelegate, TArg argument, out TBoundDelegate boundDelegate)
+            where TPooled : AbstractDelegateWithBoundArgument<TArg, TUnboundDelegate, TBoundDelegate>, new()
+            where TUnboundDelegate : Delegate
+            where TBoundDelegate : Delegate
+        {
+            var obj = DefaultDelegatePool<TPooled>.Instance.Allocate();
+            obj.Argument = argument;
+            obj.UnboundDelegate = unboundDelegate;
+            boundDelegate = obj.BoundDelegate;
+
+            return new Releaser(obj, obj => DefaultDelegatePool<TPooled>.Instance.Free((TPooled)obj));
+        }
+
+        public static Releaser GetPooledAction<TArg>(Action<TArg> unboundAction, TArg argument, out Action boundAction)
+            => GetPooledDelegate<ActionWithBoundArgument<TArg>, TArg, Action<TArg>, Action>(unboundAction, argument, out boundAction);
+
+        public static Releaser GetPooledAction<T1, TArg>(Action<T1, TArg> unboundAction, TArg argument, out Action<T1> boundAction)
+            => GetPooledDelegate<ActionWithBoundArgument<T1, TArg>, TArg, Action<T1, TArg>, Action<T1>>(unboundAction, argument, out boundAction);
+
+        public static Releaser GetPooledAction<T1, T2, TArg>(Action<T1, T2, TArg> unboundAction, TArg argument, out Action<T1, T2> boundAction)
+            => GetPooledDelegate<ActionWithBoundArgument<T1, T2, TArg>, TArg, Action<T1, T2, TArg>, Action<T1, T2>>(unboundAction, argument, out boundAction);
+
+        public static Releaser GetPooledAction<T1, T2, T3, TArg>(Action<T1, T2, T3, TArg> unboundAction, TArg argument, out Action<T1, T2, T3> boundAction)
+            => GetPooledDelegate<ActionWithBoundArgument<T1, T2, T3, TArg>, TArg, Action<T1, T2, T3, TArg>, Action<T1, T2, T3>>(unboundAction, argument, out boundAction);
+
+        public static Releaser GetPooledFunction<TArg, TResult>(Func<TArg, TResult> unboundFunction, TArg argument, out Func<TResult> boundFunction)
+            => GetPooledDelegate<FuncWithBoundArgument<TArg, TResult>, TArg, Func<TArg, TResult>, Func<TResult>>(unboundFunction, argument, out boundFunction);
+
+        public static Releaser GetPooledFunction<T1, TArg, TResult>(Func<T1, TArg, TResult> unboundFunction, TArg argument, out Func<T1, TResult> boundFunction)
```

Comment:
>GetPooledFunction [](start = 31, length = 17)

Are we using any of these `GetPooled...` methods other than this one? If not, consider removing the others, or at least remove the variants that include `T2` or `T3`.

Reply:
It is likely that additional overloads will be used as we dive deeper into performance scenarios.